coderd/provisionerdaemons.go

```diff
@@ -59,7 +62,106 @@ func (api *API) provisionerDaemons(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	httpapi.Write(rw, http.StatusOK, daemons)
+	var results []codersdk.ProvisionerDaemon
+	for _, daemon := range daemons {
+		results = append(results, convertProvisionerDaemon(daemon))
+	}
+	httpapi.Write(rw, http.StatusOK, results)
+}
+
+// Serves the provisioner daemon protobuf API over a WebSocket.
+func (api *API) provisionerDaemonsListen(rw http.ResponseWriter, r *http.Request) {
+	daemon := httpmw.ProvisionerDaemon(r)
+	api.Logger.Warn(r.Context(), "daemon connected", slog.F("daemon", daemon.Name))
+
+	api.websocketWaitMutex.Lock()
+	api.websocketWaitGroup.Add(1)
+	api.websocketWaitMutex.Unlock()
+	defer api.websocketWaitGroup.Done()
+
+	conn, err := websocket.Accept(rw, r, &websocket.AcceptOptions{
```

Comment:
Do consider that `r.Context()` is invalid from this point onwards (due to http Hijack). So if it's relied upon for connection closure / cancellation, it will not work.

You can consider using `func websocketNetConn` to rewire the context below instead of `websocket.NetConn(...)`.